File: Array-String/HashTable/CPP/tuple_with_same_product.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

static const int _ = []() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 0;
}();

class Solution {
public:
    int tupleSameProduct(vector<int>& nums) {
        unordered_map<long long, int> mult;
        int res = 0;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i+1; j < nums.size(); j++) {
                long long m = nums[i] * nums[j];
                mult[m]++;
            }
        }
        for (auto const& [k, v] : mult) {
            if (v >= 2) res += comb(v, 2) * 8;
        }
        return res;
    }

    long long comb(int n, int k) {
        if (k > n - k) k = n - k;
        long long res = 1;
        for (int i = 0; i < k; ++i) {
            res = res * (n - i) / (i + 1);
        }
        return res;
    }

    // without using comb
    int tupleSameProduct(vector<int>& nums) {
        unordered_map<long long, int> mult;
        int res = 0;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i+1; j < nums.size(); j++) {
                long long m = nums[i] * nums[j];
                mult[m]++;
            }
        }
        for (auto const& [k, v] : mult) {
            if (v >= 2) res += v * (v-1) * 4;
        }
        return res;
    }
};

File: Array-String/HashTable/Python/tuple_with_same_product.py
import math
from collections import defaultdict

class Solution:
    def factorial(self, n):
        dp = [0] * (n+1)
        dp[0] = 1
        for i in range(1, n+1):
            dp[i] = i * dp[i-1]
        return dp[n]

    def combination_product(self, n, k):
        "n len(array), k sublist to select"
        if n < k: return 0
        return self.factorial(n) // (self.factorial(k) * self.factorial(n - k))

    def tupleSameProduct(self, nums: list[int]) -> int:
        prod, ans = defaultdict(int), 0
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                mult = nums[i] * nums[j]
                prod[mult] += 1
        for k in prod:
            if prod[k] >= 2: ans += self.combination_product(prod[k]) * 8
        return ans


    # shorter version, using built-in funcs
    def tupleSameProduct(self, nums: list[int]) -> int:
        prod, ans = defaultdict(int), 0
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                mult = nums[i] * nums[j]
                prod[mult] += 1
        return sum(math.comb(v, 2) * 8 for v in prod.values() if v >= 2)